sql/analyzer/rules.go

```diff
@@ -264,6 +265,103 @@ func resolveTables(ctx *sql.Context, a *Analyzer, n sql.Node) (sql.Node, error)
 	})
 }
 
+func resolveNaturalJoins(ctx *sql.Context, a *Analyzer, n sql.Node) (sql.Node, error) {
+	span, ctx := ctx.Span("resolve_natural_joins")
+	defer span.Finish()
+
+	if n.Resolved() {
+		return n, nil
+	}
+
+	a.Log("resolving natural joins, node of type %T", n)
+	return n.TransformUp(func(n sql.Node) (sql.Node, error) {
+		a.Log("transforming node of type: %T", n)
+		if n.Resolved() {
+			return n, nil
+		}
+
+		join, ok := n.(*plan.NaturalJoin)
+		if !ok {
+			return n, nil
+		}
+
+		// we need both leaves resolved before resolving this one
+		if !join.Left.Resolved() || !join.Right.Resolved() {
+			return n, nil
+		}
+
+		leftSchema, rightSchema := join.Left.Schema(), join.Right.Schema()
+
+		var conditions, common, left, right []sql.Expression
+		var seen = make(map[string]struct{})
+
+		for i, lcol := range leftSchema {
```

Comment:
This is just a proposal to consider:
Maybe instead O(n^2) iteration here, we can:
- go through the left and store everything in `map[column_name] = expression.NewGetFieldWithTable`
- go through the right and do the same as for left but if `key/column_name` exists then overwrite it.
at the end map should contain all dedup columns in O(2*n), but maybe I'm wrong and missed something, so  it's up to you.